### ISSUE TITLE
 Is necessary removeObserver in `NSNotificationCenter` for iOS 9.0 and later or macOS 10.11 and later ？

### ISSUE THREAD

**Issue by kinarob** [rs/SDWebImage#2487]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos 
 Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
 SDWebImage Version      | e.g. 4.2.0 / 4.1.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 9 / Xcode 8
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

According to apple's documentation, 
> If your app targets iOS 9.0 and later or macOS 10.11 and later, you don't need to unregister an observer in its dealloc method. Otherwise, you should call removeObserver:name:object: before observer or any object passed to this method is deallocated.

so, `removeObserver` is necessary with iOS 9.0 and later or macOS 10.11 and later ？

Thanks 
kinarob

**Comment by dreampiggy** [comment]
The comment answer your question:

`you don't need to unregister an observer in its dealloc method`

You can still call `removeObserver` any times , even you didn't register any observer for the target notification name. It does not cause any issue because Apple implementation check whether they have current object registered in the notification maps.

**Comment by kinarob** [comment]
The table of observations contains weak pointers which are zeroed when the observers get garbage collected. so，The object will release quickly if you don't call `removeObserver`?

**Comment by zhongwuzw** [comment]
@kinarob If you target after `iOS9.0`, you don't need remove observer explicitly. otherwise, you still need it. The reason why it don't needed anymore I think is `Apple` remove the `unsafe_unretained` in `Observer` and uses `weak` instead. So no crash anymore, pointer is safe.

**Comment by kinarob** [comment]
@zhongwuzw Hi brother，I'm testing on real device（iOS 8.1.3) today, and if you forget to `removeObserver`, it still won't cause crash.

**Comment by zhongwuzw** [comment]
@kinarob Yeah, definitely, it not uses `weak`, it uses `unsafe_unretained`. Like my comment said above, your `App`'s minimun target should greater than `iOS9.0`.

**Comment by zhongwuzw** [comment]
I think this `Issue` can close. @kinarob Thanks for this discussion.

**Comment by kinarob** [comment]
Thank you for your reply, I'm so dumb. What we can be sure is that iOS 9 or above devices don't cause crash. I'm curious why we don't remove observer and still don't cause crash on iOS 8 devices. I think if `unsafe_unretained` is on iOS 8 devices, it will cause crash, but in fact It didn't cause crash, so I created this issues, guessing that `unsafe_unretained` was not used on iOS 8 devices, and that there would be no need to remove the observer later. The following is the source of research   
`- postNotificationName:object: ` [GUN  Foundation](http://ftpmain.gnustep.org/pub/gnustep/core/gnustep-base-1.25.1.tar.gz) (for reference only) .


![img_0550](https://user-images.githubusercontent.com/29303030/45523833-d9979900-b7fd-11e8-9c60-1821280cc4ff.PNG)


```javascript
/**
 * Private method to perform the actual posting of a notification.
 * Release the notification before returning, or before we raise
 * any exception ... to avoid leaks.
 */
- (void)_postAndRelease:(NSNotification*)notification
{
  Observation	*o;
  unsigned	count;
  NSString	*name = [notification name];
  id		object;
  GSIMapNode	n;
  GSIMapTable	m;
  GSIArrayItem	i[64];
  GSIArray_t	b;
  GSIArray	a = &b;

  if (name == nil) {
      RELEASE(notification);
      [NSException raise: NSInvalidArgumentException
		  format: @"Tried to post a notification with no name."];
    }
  object = [notification object];

  /*
   * Lock the table of observations while we traverse it.
   *
   * The table of observations contains weak pointers which are zeroed when
   * the observers get garbage collected.  So to avoid consistency problems
   * we disable gc while we copy all the observations we are interested in.
   * We use scanned memory in the array in the case where there are more
   * than the 64 observers we allowed room for on the stack.
   */
  GSIArrayInitWithZoneAndStaticCapacity(a, _zone, 64, i);
  lockNCTable(TABLE);

  /*
   * Find all the observers that specified neither NAME nor OBJECT.
   */
  for (o = WILDCARD = purgeCollected(WILDCARD); o != ENDOBS; o = o->next) {
      GSIArrayAddItem(a, (GSIArrayItem)o);
    }

  /*
   * Find the observers that specified OBJECT, but didn't specify NAME.
   */
  if (object) {
      n = GSIMapNodeForSimpleKey(NAMELESS, (GSIMapKey)object);
      if (n != 0) {
          o = purgeCollectedFromMapNode(NAMELESS, n);
          while (o != ENDOBS) {
              GSIArrayAddItem(a, (GSIArrayItem)o);
              o = o->next;
          }
      }
  }

  /*
   * Find the observers of NAME, except those observers with a non-nil OBJECT
   * that doesn't match the notification's OBJECT).
   */
  if (name) {
      n = GSIMapNodeForKey(NAMED, (GSIMapKey)((id)name));
      if (n){
          m = (GSIMapTable)n->value.ptr;
      } else {
          m = 0;
      }
      if (m != 0) {
	  /*
	   * First, observers with a matching object.
	   */
	  n = GSIMapNodeForSimpleKey(m, (GSIMapKey)object);
	  if (n != 0) {
	      o = purgeCollectedFromMapNode(m, n);
	      while (o != ENDOBS) {
		  GSIArrayAddItem(a, (GSIArrayItem)o);
		  o = o->next;
          }
      }

	  if (object != nil) {
	      /*
	       * Now observers with a nil object.
	       */
	      n = GSIMapNodeForSimpleKey(m, (GSIMapKey)nil);
	      if (n != 0) {
	          o = purgeCollectedFromMapNode(m, n);
              while (o != ENDOBS) {
                  GSIArrayAddItem(a, (GSIArrayItem)o);
                  o = o->next;
              }
          }
	    }
      }
    }

  /* Finished with the table ... we can unlock it,
   */
  unlockNCTable(TABLE);

  /*
   * Now send all the notifications.
   */
  count = GSIArrayCount(a);
  while (count-- > 0) {
      o = GSIArrayItemAtIndex(a, count).ext;
      if (o->next != 0) {
          NS_DURING {
              [o->observer performSelector: o->selector
                                withObject: notification];
          }
          NS_HANDLER {
              NSLog(@"Problem posting notification: %@", localException);
          }
          NS_ENDHANDLER
      }
  }
  lockNCTable(TABLE);
  GSIArrayEmpty(a);
  unlockNCTable(TABLE);

  RELEASE(notification);
}
```

**Comment by zhongwuzw** [comment]
@kinarob 👍 for your curiosity. But if you wants to get how framework implements in `iOS8`, I think you can check the assembly and `Xcode` to debug. We'd better to follow `Apple`'s documents, maybe some versions in series of `iOS8` already support `weak` behavior, but anyway, always add `removeObserver` explicitly is more robust.

I said `unsafe_unretained`, I mean framework just keep the pointer of the observer, so framework can use `unsafe_unretained`、`assign`、`address pointer` to store, these actually do the same thing.

And the source you linked is not the source code of `Apple`, it's just `GNU` project. So if you curious, you can debug and see how mechanism works, post here that we can all learned.

**Comment by kinarob** [comment]
😂😂，You're right. I think I should check the assembly and Xcode to debug，Thank you @zhongwuzw
